### PR TITLE
modify podhandling and webhook functionality

### PR DESCRIPTION
**Pull request by balintTobik**

### CODE REVIEW

**Comment by TimoLindqvist** [comment]
What is the background of this change ? There is no related issue so some description is needed why the change is necessary.

**Comment by Levovar** [comment]
the initial expectation of "status running == cgroups are created" has proven to be wrong during some 5G trial runs
firstly, application might add a readyness check for their proper cgroup config. in which case we will never be ready cause we havent yet prepared the cgroups, but we never prepare the cgroups unless we are ready

but even in addition to this, we have seen some ready / not ready state fluctuation happening regardless the presence of init containers, or readyness probes.
simply speaking K8s maintained state is not a reliable indication

instead, we should immediately and automatically try to adjust the cgroups when a Pod comes up. to avoid unnecessarily re-process an already adjusted Pod, we will set an indication ourselves into the Pod annotation saying "I was already here"

File: cmd/webhook/webhook.go [hunk 1]
```diff
@@ -142,6 +142,11 @@ func patchContainerEnv(poolRequests poolRequestMap, envPatched bool, patchList [
 	} else {
 		poolStr = types.DefaultPoolID
 	}
+	for _, env := range c.Env{
```
Comment (TimoLindqvist) [inline cmd/webhook/webhook.go:145]: Should this be first thing to do in this function ?

File: cmd/webhook/webhook.go [hunk 2]
```diff
@@ -160,7 +165,11 @@ func patchContainerForPinning(cpuAnnotation types.CPUAnnotation, patchList []pat
 	var patchItem patch
 
 	glog.V(2).Infof("Adding CPU pinning patches")
-
+	for _, volumeMount := range c.VolumeMounts{
```
Comment (TimoLindqvist) [inline cmd/webhook/webhook.go:168]: Log should be written only if we proceed to patching so the logging should be after this.

File: cmd/webhook/webhook.go [hunk 3]
```diff
@@ -321,7 +338,15 @@ func mutatePods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
 	}
 	// Add volumes if any container was patched for pinning
 	if pinningPatchAdded {
-		patchList = patchVolumesForPinning(patchList)
+		alreadyPatched := false
+		for _, volume := range pod.Spec.Volumes{
```
Comment (TimoLindqvist) [inline cmd/webhook/webhook.go:342]: Would it be enough to check only this in the beginning of this function and remove all the other checks ?

Comment (balintTobik) [reply cmd/webhook/webhook.go]: You're right! I moved a check in the begining.

File: cmd/webhook/webhook.go [hunk 4]
```diff
@@ -253,6 +252,16 @@ func mutatePods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
 
 	reviewResponse.Allowed = true
 
+	for _, volume := range pod.Spec.Volumes{
+		if volume.Name == "podinfo" || volume.Name == "hostbin" {
```
Comment (Levovar) [inline cmd/webhook/webhook.go:256]: what if somebody named his own volume podinfo or hostbin? :)
we could employ label / annotation mechanism for this purpose as well

File: pkg/sethandler/controller.go [hunk 5]
```diff
@@ -115,23 +118,28 @@ func (setHandler *SetHandler) PodChanged(oldPod, newPod v1.Pod) {
...
 		return false
 	}
 	//If the Pod is not running, its containers haven't been created yet - no cpuset cgroup is present to be adjusted by the CPUSetter
-	if pod.Status.Phase != "Running" {
+	if pod.ObjectMeta.Annotations[setterAnnotationKey] != "" {
 		return false
 	}
+	for _, container := range pod.Spec.Containers {
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			if containerStatus.Name == container.Name{
```
Comment (Levovar) [inline pkg/sethandler/controller.go:132]: combine the two statements?

File: pkg/sethandler/controller.go [hunk 6]
```diff
@@ -290,21 +303,20 @@ func getInfraContainerPath(podStatus v1.PodStatus, searchPath string) string {
 	return pathToInfraContainer
 }
 
-func (setHandler *SetHandler) applyCpusetToInfraContainer(pod v1.Pod, pathToSearchContainer string) error {
+func (setHandler *SetHandler) applyCpusetToInfraContainer(podStatus v1.PodStatus, pathToSearchContainer string) error {
 	cpuset := setHandler.poolConfig.SelectPool(types.DefaultPoolID).CPUs
 	if cpuset.IsEmpty() {
 		//Nothing to set. We will leave the container running on the Kubernetes provisioned default cpuset
-		log.Println("WARNING: for some reason DEFAULT cpuset was quite empty. Cannot adjust cpuset for infra container of pod: " + string(pod.ObjectMeta.UID))
+		log.Println("WARNING: for some reason DEFAULT cpuset was quite empty. Cannot adjust cpuset for infra container")
```
Comment (Levovar) [inline pkg/sethandler/controller.go:310]: these logs are entirely useless in production without an object identifiers
either Pod name + Pod ns, or Poid UID must be included in them to have any value

File: cmd/webhook/webhook.go (hunk 4 above)
Comment (balintTobik) [reply cmd/webhook/webhook.go]: Add annotation after Volumes patching and check it at the beginning of the function. This is something you thought?

Comment (Levovar) [reply cmd/webhook/webhook.go]: yep, but add it into the same patch operation, so it doesn't generate another webhook request

Comment (TimoLindqvist) [reply cmd/webhook/webhook.go]: It looks good now.

**Pull request merged**